File: src/swingjs/plaf/JSComboBoxUI.java
package swingjs.plaf;


import javajs.util.PT;


import jsjava.beans.PropertyChangeEvent;
import jsjava.beans.PropertyChangeListener;
import jsjavax.swing.JComboBox;
import jsjavax.swing.JComponent;
import jsjavax.swing.LookAndFeel;
import swingjs.api.DOMNode;

/**
 * A simple drop-down non-editable list for now. Using the jQuery method
 *  
 * 		$(domNode).on("change", f);
 * 
 * we can run our own function here when the selection is made. 
 * 
 */

public class JSComboBoxUI extends JSLightweightUI implements PropertyChangeListener {

	public JSComboBoxUI() {
		isContainer = true;
		setDoc();
	}
	
	@Override
	public DOMNode createDOMNode() {
		if (domNode == null)
			domNode = focusNode = createDOMObject("select", id);
		populateList();
		JComboBox b = (JComboBox) c;
		b.addPropertyChangeListener(this);
		bindMouse(domNode);
		//bindKeys(domNode); // ? perhaps?
		DOMNode.addJqueryHandledEvent(this, domNode, "change");
		DOMNode.setStyles(domNode, "z-index", "" + (getZIndex(null) + 5));
		setFocusable();
    return domNode;
	}

	@Override
	public boolean handleJSEvent(Object target, int eventType, Object jQueryEvent) {
		switch (eventType) {
		case -1:
      int index = PT.parseInt("" + DOMNode.getAttr(domNode, "selectedIndex"));
      ((JComboBox) c).setSelectedIndex(index);
			break;
		}
		return true;
	}

	
	private void populateList() {
		$(domNode).empty();
		JComboBox b = (JComboBox) c;
		int n = b.getItemCount();
		int iselect = b.getSelectedIndex();
		for (int i = 0; i < n; i++) {
			String item = b.getItemAt(i).toString();
			DOMNode option = DOMNode.createElement("option", id + "_" + (++incr));
			DOMNode.setAttr(option,  "innerHTML", item);
			if (i == iselect)
				DOMNode.setAttr(option, "selected", "true");
			DOMNode.add(domNode, option);
		}		
	}

	@Override
	protected void installJSUI() {
    LookAndFeel.installColorsAndFont(jc,
        "ComboBox.background",
        "ComboBox.foreground",
        "ComboBox.font");
	}

	@Override
	protected void uninstallJSUI() {
		// TODO Auto-generated method stub
		
	}

	@Override
	public void propertyChange(PropertyChangeEvent evt) {
		System.out.println("JSComboBoxUI " + evt);
	}
	
	@Override
  public boolean contains(JComponent c, int x, int y) {
    return false; // do not accept responsibility for this one?
}

  /**
   * Set the visiblity of the popup
   */
  public void setPopupVisible( JComboBox c, boolean v ) {
  	// TODO
  }

  /**
   * Determine the visibility of the popup
   */
  public boolean isPopupVisible( JComboBox c ) {
  	return false;
  }

  /**
   * Determine whether or not the combo box itself is traversable
   */
  public boolean isFocusTraversable( JComboBox c ) {
  	return true;
  }


}

File: site/swingjs/j2s/swingjs/plaf/JSComboBoxUI.js
Clazz.declarePackage ("swingjs.plaf");
Clazz.load (["java.beans.PropertyChangeListener", "swingjs.plaf.JSLightweightUI"], "swingjs.plaf.JSComboBoxUI", ["JU.PT", "javax.swing.LookAndFeel", "swingjs.JSToolkit", "swingjs.api.DOMNode", "swingjs.plaf.JSComponentUI"], function () {
c$ = Clazz.declareType (swingjs.plaf, "JSComboBoxUI", swingjs.plaf.JSLightweightUI, java.beans.PropertyChangeListener);
Clazz.makeConstructor (c$, 
function () {
Clazz.superConstructor (this, swingjs.plaf.JSComboBoxUI, []);
this.isContainer = true;
this.setDoc ();
});
Clazz.overrideMethod (c$, "getDOMObject", 
function () {
if (this.domNode == null) this.domNode = this.focusNode = this.createDOMObject ("select", this.id, []);
this.populateList ();
var b = this.c;
b.addPropertyChangeListener (this);
this.bindMouse (this.domNode);
var f = null;
var me = this;
{
f = function(ev) {me.handleJSEvent(this.domNode, -1, ev)};
}this.$ (this.domNode).on ("change", f);
swingjs.api.DOMNode.setStyles (this.domNode, ["z-index", "" + (swingjs.JSToolkit.getZIndex (this, null) + 5)]);
this.setFocusable ();
return this.domNode;
});
Clazz.defineMethod (c$, "populateList", 
 function () {
this.$ (this.domNode).empty ();
var b = this.c;
var n = b.getItemCount ();
var iselect = b.getSelectedIndex ();
for (var i = 0; i < n; i++) {
var item = b.getItemAt (i).toString ();
var option = swingjs.api.DOMNode.createElement ("option", this.id + "_" + (++swingjs.plaf.JSComponentUI.incr));
swingjs.api.DOMNode.setAttr (option, "innerHTML", item);
if (i == iselect) swingjs.api.DOMNode.setAttr (option, "selected", "true");
swingjs.api.DOMNode.add (this.domNode, option);
}
});
Clazz.overrideMethod (c$, "installJSUI", 
function () {
javax.swing.LookAndFeel.installColorsAndFont (this.c, "ComboBox.background", "ComboBox.foreground", "ComboBox.font");
});
Clazz.overrideMethod (c$, "uninstallJSUI", 
function () {
});
Clazz.overrideMethod (c$, "propertyChange", 
function (evt) {
System.out.println ("JSComboBoxUI " + evt);
}, "java.beans.PropertyChangeEvent");
Clazz.overrideMethod (c$, "contains", 
function (c, x, y) {
return false;
}, "javax.swing.JComponent,~N,~N");
Clazz.overrideMethod (c$, "handleJSEvent", 
function (target, eventType, jQueryEvent) {
switch (eventType) {
case -1:
var index = JU.PT.parseInt ("" + swingjs.api.DOMNode.getAttr (this.domNode, "selectedIndex"));
(this.c).setSelectedIndex (index);
break;
}
return true;
}, "~O,~N,~O");
Clazz.defineMethod (c$, "setPopupVisible", 
function (c, v) {
}, "javax.swing.JComboBox,~B");
Clazz.defineMethod (c$, "isPopupVisible", 
function (c) {
return false;
}, "javax.swing.JComboBox");
Clazz.defineMethod (c$, "isFocusTraversable", 
function (c) {
return true;
}, "javax.swing.JComboBox");
});
